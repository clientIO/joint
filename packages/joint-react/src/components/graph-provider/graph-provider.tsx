import type { dia } from '@joint/core';
import type { GraphLink } from '../../types/link-types';
import {
  GraphAreElementsMeasuredContext,
  GraphStoreContext,
} from '../../context/graph-store-context';
import { useEffect, useState, type PropsWithChildren } from 'react';
import { createStore, type Store } from '../../data/create-store';
import { useElements } from '../../hooks/use-elements';
import { useGraph } from '../../hooks';
import { setLinks } from '../../utils/cell/set-cells';
import type { GraphElement } from '../../types/element-types';

interface GraphProviderHandlerProps {
  /**
   * Initial links to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly initialLinks?: Array<dia.Link | GraphLink>;
}

/**
 * GraphProviderHandler component is used to handle the graph instance and provide it to the children.
 * It also handles the default elements and links.
 * @param props - {GraphProviderHandler} props
 * @param props.children - Children to render.
 * @param props.initialLinks - Initial links to be added to graph
 * @returns GraphProviderHandler component
 * @private
 */
function GraphProviderHandler({
  children,
  initialLinks,
}: PropsWithChildren<GraphProviderHandlerProps>) {
  const areElementsMeasured = useElements((items) => {
    let areMeasured = true;
    for (const [, { width = 0, height = 0 }] of items) {
      if (width <= 1 || height <= 1) {
        areMeasured = false;
        break;
      }
    }
    return areMeasured;
  });
  const graph = useGraph();

  useEffect(() => {
    if (areElementsMeasured) {
      setLinks({ graph, initialLinks });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [areElementsMeasured, graph]);

  return (
    <GraphAreElementsMeasuredContext.Provider value={areElementsMeasured}>
      {children}
    </GraphAreElementsMeasuredContext.Provider>
  );
}

export interface GraphProps {
  /**
   * Graph instance to use. If not provided, a new graph instance will be created.
   * @see https://docs.jointjs.com/api/dia/Graph
   * @default new dia.Graph({}, { cellNamespace: shapes })
   */
  readonly graph?: dia.Graph;
  /**
   * Children to render.
   */
  readonly children?: React.ReactNode;
  /**
   * Namespace for cell models.
   * It's loaded just once, so it cannot be used as React state.
   * When added new shape, it will not remove existing ones, it will just add new ones.
   * So `{ ...shapes, ReactElement }` elements are still available.
   * @default  `{ ...shapes, ReactElement }`
   * @see https://docs.jointjs.com/api/shapes
   */
  readonly cellNamespace?: unknown;
  /**
   * Custom cell model to use.
   * It's loaded just once, so it cannot be used as React state.
   * @see https://docs.jointjs.com/api/dia/Cell
   */
  readonly cellModel?: typeof dia.Cell;
  /**
   * Initial elements to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly initialElements?: Array<dia.Element | GraphElement>;
  /**
   * Initial links to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly initialLinks?: Array<dia.Link | GraphLink>;

  /**
   * Store is build around graph, it handles react updates and states, it can be created separately and passed to the provider via `createStore` function.
   * @see `createStore`
   */
  readonly store?: Store;
}

/**
 *
 * GraphProvider component creates a graph instance and provide `dia.graph` to it's children.
 * It relies on @see useCreateGraphStore hook to create the graph instance.
 *
 * Without this provider, the library will not work.
 * @param props - {GraphProvider} props
 * @returns GraphProvider component
 * @example
 * Using provider:
 * ```tsx
 * import { GraphProvider } from '@joint/react'
 *
 * function App() {
 *  return (
 *   <GraphProvider>
 *    <MyApp />
 *  </GraphProvider>
 * )
 * ```
 * @example
 * Using provider with default elements and links:
 * ```tsx
 * import { GraphProvider } from '@joint/react'
 *
 * function App() {
 *  return (
 *   <GraphProvider initialElements={[]} initialLinks={[]}>
 *    <MyApp />
 *  </GraphProvider>
 * )
 * ```
 * @group Components
 */
export function GraphProvider(props: Readonly<GraphProps>) {
  const { children, initialLinks, store, ...rest } = props;

  /**
   * Graph store instance.
   * @returns - The graph store instance.
   */

  const [graphStore, setGraphStore] = useState<null | Store>(null);

  useEffect(() => {
    const newStore = store ?? createStore({ ...rest });
    // We must use state initialization for the store, because it can be used in the same component.
    // eslint-disable-next-line @eslint-react/hooks-extra/no-direct-set-state-in-use-effect
    setGraphStore(newStore);
    return () => {
      if (newStore) {
        newStore.destroy();
      }
    };
    // On load initialization
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  if (graphStore === null) {
    return null;
  }

  return (
    <GraphStoreContext.Provider value={graphStore}>
      <GraphProviderHandler initialLinks={initialLinks}>{children}</GraphProviderHandler>
    </GraphStoreContext.Provider>
  );
}
