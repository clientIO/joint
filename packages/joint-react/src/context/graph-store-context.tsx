import { createContext } from 'react';
import type { Store } from '../data/create-store';
export type StoreContext = Store;

export const GraphStoreContext = createContext<Store | undefined>(undefined);
export const GraphAreElementsMeasuredContext = createContext<boolean>(false);
