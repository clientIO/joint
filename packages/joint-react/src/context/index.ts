export * from './cell-id.context';
export * from './graph-store-context';
export * from './paper-context';
export * from './tools-view.context';
