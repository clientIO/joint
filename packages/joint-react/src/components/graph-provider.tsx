import { GraphContext } from '../context/graph-context'
import type { dia } from '@joint/core'
import { useCreateGraphStore } from '../hooks/use-create-graph-store'

export interface GraphProps {
  /**
   * Graph instance to use. If not provided, a new graph instance will be created.
   * @see https://docs.jointjs.com/api/dia/Graph
   * @default new dia.Graph({}, { cellNamespace: shapes })
   */
  readonly graph?: dia.Graph
  /**
   * Children to render.
   */
  readonly children: React.ReactNode
  /**
   * Namespace for cell models.
   * It's loaded just once, so it cannot be used as React state.
   * @default shapes
   * @see https://docs.jointjs.com/api/shapes
   */
  readonly cellNamespace?: unknown
  /**
   * Custom cell model to use.
   * It's loaded just once, so it cannot be used as React state.
   * @see https://docs.jointjs.com/api/dia/Cell
   */
  readonly cellModel?: typeof dia.Cell
  /**
   * Initial elements to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly defaultElements?: Array<dia.Element>

  /**
   * Initial links to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly defaultLinks?: Array<dia.Link>
}

/**
 * GraphProvider component creates a graph instance and provides it to its children via context.
 * It also handles updates to the graph when cells change via React state or JointJS events.
 */
export function GraphProvider(props: GraphProps) {
  const { children, ...rest } = props
  const graphStore = useCreateGraphStore(rest)

  return <GraphContext.Provider value={graphStore}>{children}</GraphContext.Provider>
}
