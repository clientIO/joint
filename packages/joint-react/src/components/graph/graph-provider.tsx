import type { dia } from '@joint/core';
import type { GraphLink } from '../../types/link-types';
import {
  forwardRef,
  useLayoutEffect,
  type Dispatch,
  type PropsWithChildren,
  type SetStateAction,
} from 'react';
import { createStore, type GraphStore } from '../../data/create-graph-store';
import { useElements } from '../../hooks/use-elements';
import { useGraphStore } from '../../hooks';
import type { GraphElement } from '../../types/element-types';
import { CONTROLLED_MODE_BATCH_NAME } from '../../utils/graph/update-graph';
import { useImperativeApi } from '../../hooks/use-imperative-api';
import { GraphAreElementsMeasuredContext, GraphStoreContext } from '../../context';
import { getTargetOrSource } from '../../utils/cell/get-link-targe-and-source-ids';

interface GraphProviderBaseProps<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Element extends dia.Element | GraphElement = any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Link extends dia.Link | GraphLink = any,
> {
  /**
   * Elements (nodes) to be added to graph.
   * When `onElementsChange`, it enabled controlled mode.
   * If there is no `onElementsChange` provided, it will be used just on onload (initial)
   */
  readonly elements?: Element[];

  /**
   * Links (edges) to be added to graph.
   * When `onLinksChange`, it enabled controlled mode.
   * If there is no `onLinksChange` provided, it will be used just on onload (initial)
   */
  readonly links?: Link[];

  /**
   * Callback triggered when elemeÂ§nts (nodes) change.
   * Providing this prop enables controlled mode for elements.
   * If specified, this function will override the default behavior, allowing you to manage all element changes manually instead of relying on `graph.change`.
   */
  readonly onElementsChange?: Dispatch<SetStateAction<Element[]>>;

  /**
   * Callback triggered when links (edges) change.
   * Providing this prop enables controlled mode for links.
   * If specified, this function will override the default behavior, allowing you to manage all link changes manually instead of relying on `graph.change`.
   */
  readonly onLinksChange?: Dispatch<SetStateAction<Link[]>>;
}

/**
 * Internal handler coordinating initial population and controlled-mode mirroring
 * for elements and links. Also delays link creation until elements are measured
 * in uncontrolled mode to avoid flicker.
 * @param props - The properties for the GraphProviderHandler, including elements, links, and callbacks.
 * @returns A context provider for the measured state of elements.
 * @private
 */
export function GraphProviderHandler<
  Element extends dia.Element | GraphElement = dia.Element,
  Link extends dia.Link | GraphLink = dia.Link,
>(props: PropsWithChildren<GraphProviderBaseProps<Element, Link>>) {
  const { elements, links, onElementsChange, onLinksChange, children } = props;
  const areElementsMeasured = useElements((items) => {
    let areMeasured = true;
    for (const { width = 0, height = 0 } of items) {
      if (width <= 1 || height <= 1) {
        areMeasured = false;
        break;
      }
    }
    return areMeasured;
  });

  const { graph, setElements, setLinks } = useGraphStore();

  const areElementsInControlledMode = !!onElementsChange;
  const areLinksInControlledMode = !!onLinksChange;
  const isControlledMode = areElementsInControlledMode || areLinksInControlledMode;
  // Controlled mode for elements
  useLayoutEffect(() => {
    if (!areElementsMeasured) return;
    if (!graph) return;
    if (!isControlledMode) return;

    graph.startBatch(CONTROLLED_MODE_BATCH_NAME);
    if (areElementsInControlledMode && elements !== undefined) {
      setElements(elements);
    }
    if (areLinksInControlledMode && links !== undefined) {
      setLinks(links as GraphLink[]);
    }
    graph.stopBatch(CONTROLLED_MODE_BATCH_NAME);
  }, [
    areElementsInControlledMode,
    areElementsMeasured,
    areLinksInControlledMode,
    graph,
    elements,
    links,
    isControlledMode,
    setElements,
    setLinks,
  ]);

  useLayoutEffect(() => {
    // with this all links are connected only when react elements are measured
    // It fixes issue with a flickering of un-measured react elements.
    if (isControlledMode) return;
    if (!areElementsMeasured) return;
    if (!links?.length) return;
    const hasSomePort = links?.some((link) => {
      const { source, target } = link;
      const sourceObject = getTargetOrSource(source as dia.Link.EndJSON);
      const targetObject = getTargetOrSource(target as dia.Link.EndJSON);
      return sourceObject.port || targetObject.port;
    });
    if (!hasSomePort) return;
    setLinks(links as GraphLink[]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [areElementsMeasured, isControlledMode]);

  return (
    <GraphAreElementsMeasuredContext.Provider value={areElementsMeasured}>
      {children}
    </GraphAreElementsMeasuredContext.Provider>
  );
}

export interface GraphProps<
  Graph extends dia.Graph = dia.Graph,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Element extends dia.Element | GraphElement = any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Link extends dia.Link | GraphLink = any,
> extends GraphProviderBaseProps<Element, Link> {
  /**
   * Graph instance to use. If not provided, a new graph instance will be created.
   * @see https://docs.jointjs.com/api/dia/Graph
   * @default new dia.Graph({}, { cellNamespace: shapes })
   */
  readonly graph?: Graph;
  /**
   * Children to render.
   */
  readonly children?: React.ReactNode;
  /**
   * Namespace for cell models.
   * It's loaded just once, so it cannot be used as React state.
   * When added new shape, it will not remove existing ones, it will just add new ones.
   * So `{ ...shapes, ReactElement }` elements are still available.
   * @default  `{ ...shapes, ReactElement }`
   * @see https://docs.jointjs.com/api/shapes
   */
  readonly cellNamespace?: unknown;
  /**
   * Custom cell model to use.
   * It's loaded just once, so it cannot be used as React state.
   * @see https://docs.jointjs.com/api/dia/Cell
   */
  readonly cellModel?: typeof dia.Cell;

  /**
   * Store is build around graph, it handles react updates and states, it can be created separately and passed to the provider via `createStore` function.
   * @see `createStore`
   */
  readonly store?: GraphStore<Graph>;
}

/**
 * Graph component creates a graph instance and provides `dia.graph` to its children.
 * This component is essential for the library to function correctly. It manages the graph instance and supports controlled and uncontrolled modes for elements and links.
 * @param props - The properties for the Graph component.
 * @param forwardedRef - A reference to the GraphStore instance.
 * @returns The Graph component.
 * @example
 * Using the Graph component:
 * ```tsx
 * import { Graph } from '@joint/react';
 * function App() {
 *   return (
 *     <Graph>
 *       <MyApp />
 *     </Graph>
 *   );
 * }
 * ```
 * @example
 * Using the Graph component with default elements and links:
 * ```tsx
 * import { Graph } from '@joint/react';
 * function App() {
 *   return (
 *     <Graph elements={[]} links={[]}>
 *       <MyApp />
 *     </Graph>
 *   );
 * }
 * ```
 */
function GraphBase<Element extends dia.Element | GraphElement, Link extends dia.Link | GraphLink>(
  props: Readonly<GraphProps<dia.Graph, Element, Link>>,
  forwardedRef: React.Ref<GraphStore>
) {
  const { children, store, ...rest } = props;
  /**
   * Graph store instance.
   * @returns - The graph store instance.
   */

  const { isReady, ref } = useImperativeApi<GraphStore>(
    {
      forwardedRef,
      onLoad() {
        const newStore = store ?? createStore({ ...rest });
        // We must use state initialization for the store, because it can be used in the same component.

        return {
          cleanup() {
            if (newStore) {
              newStore.destroy(!!rest.graph || !!store?.graph);
            }
          },
          instance: newStore,
        };
      },
    },
    []
  );

  if (!isReady) {
    return null;
  }

  return (
    <GraphStoreContext.Provider value={ref.current}>
      <GraphProviderHandler {...props}>{children}</GraphProviderHandler>
    </GraphStoreContext.Provider>
  );
}

/**
 * GraphProviderHandler component is used to handle the graph instance and provide it to the children.
 * It also handles the default elements and links.
 * @returns GraphProviderHandler component
 * @param props - {GraphProviderHandler} props
 * @private
 */
export const GraphProvider = forwardRef(GraphBase) as <
  Element extends dia.Element | GraphElement = dia.Element,
  Link extends dia.Link | GraphLink = dia.Link,
>(
  props: Readonly<GraphProps<dia.Graph, Element, Link>> & {
    ref?: React.Ref<GraphStore>;
  }
) => ReturnType<typeof GraphBase>;
