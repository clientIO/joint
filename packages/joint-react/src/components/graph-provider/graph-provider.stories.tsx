/* eslint-disable sonarjs/no-unused-vars */
/* eslint-disable sonarjs/pseudo-random */
/* eslint-disable @eslint-react/dom/no-missing-button-type */
/* eslint-disable react-perf/jsx-no-new-function-as-prop */

import type { Meta, StoryObj } from '@storybook/react/*';
import { GraphProvider } from './graph-provider';
import { createElements, createLinks, type InferElement, ReactElement } from '@joint/react';
import { Paper, type RenderElement } from '../paper/paper';
import { dia } from '@joint/core';
import { BUTTON_CLASSNAME, PAPER_CLASSNAME, PRIMARY } from 'storybook-config/theme';
import { makeRootDocumentation, makeStory } from '@joint/react/src/stories/utils/make-story';
import { getAPILink } from '@joint/react/src/stories/utils/get-api-documentation-link';
import { HTMLNode } from 'storybook-config/decorators/with-simple-data';

const API_URL = getAPILink('GraphProvider');

export type Story = StoryObj<typeof GraphProvider>;
const meta: Meta<typeof GraphProvider> = {
  title: 'Components/GraphProvider',
  component: GraphProvider,
  parameters: makeRootDocumentation({
    description: `
GraphProvider is a component that provides a graph context to its children. It is used to manage and render graph elements.
    `,
    apiURL: API_URL,
    code: `import { GraphProvider } from '@joint/react'
<GraphProvider>
  <Paper renderElement={({width, height}) => <rect rx={10} ry={10} width={width} height={height} fill={"blue"} />} />
</GraphProvider>
    `,
  }),
};

export default meta;

const STYLE = { padding: 10, backgroundColor: PRIMARY, borderRadius: 10, width: 80 };

const initialElementsWithSize = createElements([
  { id: 1, width: 100, height: 50, x: 20, y: 200, color: PRIMARY },
  { id: 2, width: 100, height: 50, x: 200, y: 200, color: PRIMARY },
]);
const initialElementsWithoutSize = createElements([
  { id: 1, x: 20, y: 200, color: PRIMARY },
  { id: 2, x: 200, y: 200, color: PRIMARY },
]);
const initialLinks = createLinks([
  {
    id: '1-1',
    source: 2,
    target: 1,
    attrs: {
      line: {
        stroke: PRIMARY,
      },
    },
  },
]);

type ElementType = InferElement<typeof initialElementsWithSize>;

function RenderElement({ color, width, height }: ElementType) {
  return <rect rx={10} ry={10} className="node" width={width} height={height} fill={color} />;
}
function PaperChildren(props: Readonly<{ renderElement?: RenderElement<ElementType> }>) {
  const { renderElement = RenderElement } = props;
  return <Paper className={PAPER_CLASSNAME} renderElement={renderElement} width={'100%'} />;
}

export const Default = makeStory<Story>({
  args: {
    initialElements: initialElementsWithSize,
    children: <PaperChildren />,
  },

  apiURL: API_URL,
  description: 'Default graph provider with rectangle children.',
});

export const WithExternalGraph = makeStory<Story>({
  args: {
    initialElements: initialElementsWithSize,
    children: <PaperChildren />,
    graph: new dia.Graph({}, { cellNamespace: { ReactElement } }),
  },

  apiURL: API_URL,
  description: 'Graph provider with external graph.',
  code: `import { GraphProvider } from '@joint/react'
import { dia } from '@joint/core';
import { Paper } from '../paper/paper';
import { ReactElement } from '@joint/react/src/core/react-element';
const graph = new dia.Graph({}, { cellNamespace: { ReactElement } });
<GraphProvider graph={graph}>
  <Paper renderElement={({width, height}) => <rect rx={10} ry={10} width={width} height={height} fill={"blue"} />} />
</GraphProvider>
  `,
});

export const WithLink = makeStory<Story>({
  args: {
    initialLinks,
    initialElements: initialElementsWithSize,
    children: <PaperChildren />,
  },

  apiURL: API_URL,
  description: 'Graph provider with links.',
});

export const WithoutSizeDefinedInElements = makeStory<Story>({
  args: {
    initialLinks,
    initialElements: initialElementsWithoutSize,
    children: (
      <PaperChildren renderElement={() => <HTMLNode style={STYLE}>Hello world!</HTMLNode>} />
    ),
  },

  apiURL: API_URL,
  description: 'Graph provider without size defined in elements.',
});

const graph = new dia.Graph({}, { cellNamespace: { ReactElement } });

function generateRandomElements(length: number) {
  return createElements(
    Array.from({ length }, (_, index) => ({
      id: `node-${index}`,
      width: 100,
      height: 50,
      x: Math.random() * 500,
      y: Math.random() * 500,
      color: 'magenta',
    }))
  );
}

export const WithExternalGraphAndLayout = makeStory<Story>({
  args: {
    graph,
    initialElements: generateRandomElements(20),
    children: (
      <>
        <button
          className={BUTTON_CLASSNAME}
          onClick={() => {
            const elements = graph.getCells(); // Get all elements in the graph
            const rowWidth = 500; // Define the maximum width of each row
            let currentX = 0;
            let currentY = 0;
            let rowWidthUsed = 0;

            for (const [_, element] of elements.entries()) {
              const elementWidth = 100; // Set width for the element (you can use element.getBBox().width if dynamic)
              if (rowWidthUsed + elementWidth > rowWidth) {
                // Move to the next row
                currentX = 0;
                currentY += 85; // Add some vertical space between rows
                rowWidthUsed = 0;
              }
              if (!element.isElement()) {
                continue;
              }

              // Set the new position for the element
              element.position(currentX, currentY);

              // Update the current X and row width used
              currentX += elementWidth;
              rowWidthUsed += elementWidth;
            }
          }}
        >
          Make layout
        </button>
        <PaperChildren renderElement={() => <HTMLNode style={STYLE}>Hello world!</HTMLNode>} />
      </>
    ),
  },

  apiURL: API_URL,
  description: 'Graph provider with external graph and layout.',
  code: `import { GraphProvider } from '@joint/react'
import { dia } from '@joint/core';
import { Paper } from '../paper/paper';
import { ReactElement } from '@joint/react/src/core/react-element';
import { DirectedGraph } from '@joint/layout-directed-graph';
const graph = new dia.Graph({}, { cellNamespace: { ReactElement } });
const elements = generateRandomElements(20);
<GraphProvider graph={graph} initialElements={elements}>
  <button
    onClick={() => {
      DirectedGraph.layout(graph, {
        setLinkVertices: true,
        marginX: 2,
        marginY: 2,
        align: 'DR',
      });
    }
  >
    Make layout
  </button>
  <Paper renderElement={({ width, height }) => <rect rx={10} ry={10} width={width} height={height} fill={"blue"} />} />
</GraphProvider>
  `,
});
