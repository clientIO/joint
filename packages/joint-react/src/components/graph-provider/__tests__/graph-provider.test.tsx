import React from 'react';
import { act, render, waitFor } from '@testing-library/react';
import { GraphStoreContext } from '../../../context/graph-store-context';
import { GraphProvider } from '../graph-provider';
import { createStore, type Store } from '../../../data/create-store';
import { dia } from '@joint/core';
import { useElements, useLinks } from '../../../hooks';
import { createElements } from '../../../utils/create';
import * as stories from '../graph-provider.stories';
import { runStorybookSnapshot } from '../../../utils/run-storybook-snapshot';

runStorybookSnapshot({
  Component: GraphProvider,
  stories,
  name: 'GraphProvider',
});
describe('graph-provider', () => {
  it('should render children and match snapshot', () => {
    const { asFragment, getByText } = render(
      <GraphProvider>
        <div>Child Content</div>
      </GraphProvider>
    );
    expect(getByText('Child Content')).toMatchSnapshot();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should provide a graph instance in context', () => {
    let contextGraph: Store | undefined;
    function TestComponent() {
      contextGraph = React.useContext(GraphStoreContext);
      return null;
    }
    render(
      <GraphProvider>
        <TestComponent />
      </GraphProvider>
    );
    expect(contextGraph).toBeInstanceOf(Object);
  });

  it('should render graph provider with links and elements', async () => {
    const elements = createElements([
      {
        width: 100,
        height: 100,
        id: 'element1',
      },
    ]);
    const link = new dia.Link({ id: 'link1', type: 'standard.Link', source: { id: 'element1' } });
    let linkCount = 0;
    let elementCount = 0;
    function TestComponent() {
      linkCount = useElements((items) => items.size);
      elementCount = useLinks((items) => {
        return items.size;
      });
      return null;
    }
    render(
      // eslint-disable-next-line react-perf/jsx-no-new-array-as-prop
      <GraphProvider initialElements={elements} initialLinks={[link]}>
        <TestComponent />
      </GraphProvider>
    );

    await waitFor(() => {
      expect(linkCount).toBe(1);
      expect(elementCount).toBe(1);
    });
  });
  it('should add elements and links after initial load and useElements and useLinks should catch them', async () => {
    const graph = new dia.Graph();
    let linkCount = 0;
    let elementCount = 0;
    // eslint-disable-next-line sonarjs/no-identical-functions
    function TestComponent() {
      linkCount = useElements((items) => items.size);
      elementCount = useLinks((items) => {
        return items.size;
      });
      return null;
    }
    render(
      <GraphProvider graph={graph}>
        <TestComponent />
      </GraphProvider>
    );

    await waitFor(() => {
      expect(linkCount).toBe(0);
      expect(elementCount).toBe(0);
    });

    act(() => {
      graph.addCells([
        new dia.Element({ id: 'element1', type: 'standard.Rectangle' }),
        new dia.Link({ id: 'link1', type: 'standard.Link', source: { id: 'element1' } }),
      ]);
    });

    await waitFor(() => {
      expect(linkCount).toBe(1);
      expect(elementCount).toBe(1);
    });
  });

  it('should initialize with default elements', async () => {
    const elements = createElements([
      { width: 100, height: 100, id: 'element1' },
      { width: 200, height: 200, id: 'element2' },
    ]);
    let elementCount = 0;
    function TestComponent() {
      elementCount = useElements((items) => items.size);
      return null;
    }
    render(
      <GraphProvider initialElements={elements}>
        <TestComponent />
      </GraphProvider>
    );

    await waitFor(() => {
      expect(elementCount).toBe(2);
    });
  });

  it('should use provided store and clean up on unmount', () => {
    const mockDestroy = jest.fn();
    const mockStore = createStore({});
    // @ts-expect-error its just unit test, readonly is not needed
    mockStore.destroy = mockDestroy;

    const { unmount } = render(
      <GraphProvider store={mockStore}>
        <div>Test</div>
      </GraphProvider>
    );

    expect(mockDestroy).not.toHaveBeenCalled();
    unmount();
    expect(mockDestroy).toHaveBeenCalled();
  });
});
