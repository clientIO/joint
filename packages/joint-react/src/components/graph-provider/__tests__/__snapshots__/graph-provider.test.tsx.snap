// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphProvider "Default": GraphProvider-Default 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`GraphProvider "WithExternalGraph": GraphProvider-WithExternalGraph 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`GraphProvider "WithExternalGraphAndLayout": GraphProvider-WithExternalGraphAndLayout 1`] = `"<button class="bg-blue-500 cursor-pointer hover:bg-blue-700 text-white font-bold py-2 px-4 rounded m-2 text-sm">Make layout</button><div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`GraphProvider "WithLink": GraphProvider-WithLink 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 1; pointer-events: all; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`GraphProvider "WithoutSizeDefinedInElements": GraphProvider-WithoutSizeDefinedInElements 1`] = `"<div class="border-1 border-gray-300 rounded-lg shadow-md overflow-hidden p-2 mr-2" style="opacity: 0; pointer-events: none; position: relative; overflow: hidden; width: 100%; height: 100%;"><div class="joint-border-1 joint-border-gray-300 joint-rounded-lg joint-shadow-md joint-overflow-hidden joint-p-2 joint-mr-2 joint-theme-default" style="position: relative; width: 100%; height: 600px;"><div style="position: absolute; inset: 0;" class="joint-paper-background" joint-selector="background"></div><svg width="100%" height="100%" xmlns:xlink="http://www.w3.org/1999/xlink" style="position: absolute; inset: 0; overflow: hidden;" joint-selector="svg" id="v-*"><style id="v-*" type="text/css"></style><defs joint-selector="defs"></defs><g class="joint-layers" joint-selector="layers"><g style="pointer-events: none;" class="joint-grid-layer"></g><g class="joint-back-layer"></g><g class="joint-cells-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-labels-layer joint-viewport" style="webkit-user-select: none; user-select: none;"></g><g class="joint-front-layer"></g><g class="joint-tools-layer"></g></g></svg></div></div>"`;

exports[`graph-provider should render children and match snapshot 1`] = `
<div>
  Child Content
</div>
`;

exports[`graph-provider should render children and match snapshot 2`] = `
<DocumentFragment>
  <div>
    Child Content
  </div>
</DocumentFragment>
`;
